Use Dropbox's zxcvbn package to calculate password strength

Lusitaniae opened the pull request:
Previous password strength was solely base on password length. 

It's now using Dropbox's library to calculate password strength.

More info: https://www.npmjs.com/package/zxcvbn

> More secure: policies often fail both ways, allowing weak passwords (P@ssword1) and disallowing strong passwords.

Fixes #12910.

Lusitaniae committed:
Use Dropbox's zxcvbn package to calculate password strength

Lusitaniae committed:
Use Dropbox's zxcvbn package to calculate password strength

Lusitaniae committed:
Update password rating classification

Lusitaniae committed:
Update tests

Lusitaniae committed:
Update tests

torkelo commented:
Thanks for contributing, we will try to review and merge this soon

Limess commented:
It's worth noting that `zxcvbn.js` is a huge `400kB` gzipped. I'd highly recommend lazy loading this when used or loading it asynchronously on only the registration pages.

Commit:
Dynamically import zxcvbn

zxcvbn is 400KB gzipped. This is pretty significant.

Use the old length check and lazy load zxcvbn. Use zxcvbn once it is loaded

Commit:
Use PureComponent for PasswordStrength

This avoids recalculating the strength with zxcvbn if not necessary

Lusitaniae commented:
Thanks @Limess,

I've merged your PR to my branch

Limess commented:
Hey @Lusitaniae - sorry about this! I believe changing the imports in 
to

```jsx
import React, {PureComponent} from 'react';
```

and removing the second line `zxcvbn` import which is now dynamic.

will fix up the tests and linting errors.

Limess commented:
Looks like there's just something which isn't prettified correctly with [prettier](https://prettier.io/).

Running `npx prettier --write "**/*.{ts,tsx,scss}"` should solve that.

Lusitaniae commented:
Would have seen it if I hadn't made the change from GitHub :laughing:

torkelo commented on public/app/core/components/PasswordStrength.tsx:
```diff
@@ -1,36 +1,60 @@
     super(props);
+
+    this.state = {
+      getScore(password: string) {
```
This PR looks good except this, defining a function in the state object looks very unorthodox, have never seen it before. Also State is untyped so if you want to use state it should have a type. @dprokop have you seen this before? adding a function to the state?

Limess replied:
I don't think this is particularly unorthodox - state, like props, is just a regular collection of whatever you want, functions not being special - the alternative I can see is to lazy load the whole component using zxcvbn and replace that, passing `getScore` as a prop, but that's just moving the logic upwards to the consumer.

dprokop replied:
I haven't seen this approach being used neither. Not sure what the benefit of having this method in state is compared to regular instance method, bu this feels flaky. I see this being driven only by the dynamic import in `componentDidMount`. For that I suggest creating a higher order component that would serve as module loader and deliver the loaded module or part of its API to the desired component (PasswordStrength in this case)

torkelo commented:
We have discussed this and not sure a password strength is needed for Grafana, most use external auth anyway and having this strength value is unnecessary. 

Improving this strength value means we should enforce it in the backend so we should really use a backend lib for this. And then we need to add a setting so admins can define their own level of strength etc. 

We decided to instead remove this strength value: https://github.com/grafana/grafana/issues/17748
